app/controllers/api/v1/lease_up_applications_controller.rb

```diff
@@ -3,24 +3,57 @@
 module Api::V1
   # Lease Up Applications controller for access via the API
   class LeaseUpApplicationsController < ApiController
+    OFFSET_LIMIT = 2000
+
     def index
       if listing_type == Force::Listing::LISTING_TYPE_FIRST_COME_FIRST_SERVED
         # First Come, First Served listings don't have preferences
         applications = soql_lease_up_application_service.lease_up_applications(lease_up_apps_params)
         applications[:records] = Force::Application.convert_list(applications[:records], :from_salesforce, :to_domain)
+
+        applications[:listing_type] = listing_type
+        render json: applications
       else
         # All other listings need to be queried by preferences first
-        applications = soql_preference_service.app_preferences_for_listing(lease_up_apps_params)
-        applications[:records] = Force::Preference.convert_list(applications[:records], :from_salesforce, :to_domain)
+        application_preferences = get_application_preferences
+        general_applications = get_application_preferences(true)
+        combined = [*application_preferences[:applications], *general_applications[:applications]]
+        records = Force::Preference.convert_list(combined, :from_salesforce, :to_domain)
+        total_size = application_preferences[:acc_size] + general_applications[:acc_size]
+        page = '0'
+        pages = total_size / 10_000.to_f
+        render json: {
+          records: records,
+          total_size: total_size,
+          page: page,
+          pages: pages,
+          listing_type: listing_type,
+        }
       end
-
-      # providing the listing type so we know how to handle the response
-      applications[:listing_type] = listing_type
-      render json: applications
     end
 
     private
 
+    def get_application_preferences(general = false, prev_applications = nil, acc_size = 0) # rubocop:disable Style/OptionalBooleanParameter
+      params = lease_up_apps_params
+      if prev_applications
+        if general
+          params[:general_lottery_rank] = prev_applications[-1]['Application']['General_Lottery_Rank']
+        else
+          params[:preference_order] = prev_applications[-1]['Preference_Order']
+          params[:preference_lottery_rank] = prev_applications[-1]['Preference_Lottery_Rank']
+        end
+      end
+
+      applications_response = soql_preference_service.app_preferences_for_listing(params, general)
+      applications = [*prev_applications, *applications_response[:records]]
+      acc_size += applications_response.total_size
+
+      return { applications: applications, acc_size: acc_size } unless applications_response.total_size > OFFSET_LIMIT
+
+      get_application_preferences(general, applications, acc_size)
```

Comment:
suggestion (non-blocking): Try to avoid recursive method calls, ruby has other ways to loop through logic like blocks and lambdas.

Reply:
I think this is a perfect use case for recursion because it's necessary here that the next call depends on the response of the previous call - let's discuss?

Reply:
Happy to talk about it, here's an alternative implementation without recursion:
```ruby
def foo(general: false)
  params = lease_up_apps_params
  total_size = 0
  applications = []
  failsafe_count = 1_000

  failsafe_count.times do
    if applications.empty? && general
      params[:general_lottery_rank] = prev_applications[-1]['Application']['General_Lottery_Rank']
    elsif applications.empty?
      params[:preference_order] = applications[-1]['Preference_Order']
      params[:preference_lottery_rank] = applications[-1]['Preference_Lottery_Rank']
    end
    applications_response = soql_preference_service.app_preferences_for_listing(params, general)
    applications.concat(applications_response[:records])
    return [applications, total_size] unless applications_response.total_size > OFFSET_LIMIT
  end

  raise 'escaped infinite loop'
end
```

Reply:
this led me down a cool rabbit hole!! Found this: https://nithinbekal.com/posts/ruby-tco/

Didn't know about tail recursion, that they are the same as loops, impact on stack size, etc